### PR TITLE
[WIP] Add exact Hessian

### PR DESCRIPTION
**Pull request by giaf**
- add sim sens_adj & sens_hess to matlab interface and fix/clean them in C
- add compute_hess to dynamics and constraints modules

### CODE REVIEW

File: acados/sim/sim_irk_integrator.c [hunk 1]
```diff
@@ -1176,12 +1195,12 @@ int sim_irk(void *config_, sim_in *in, sim_out *out, void *opts_, void *mem_, vo
                         }
                     }
                     // dk_dw0
-                    blasfeo_dgecpsc(nx, nx+nu, -1.0, dK_dxu_ss, ii * nx, 0, &dxkzu_dw0, nx, 0);
+                    blasfeo_dgecpsc(nx, nx+nu, -1.0, dK_dxu_ss, ii*nx, 0, &dxkzu_dw0, nx, 0);
                     // dz_dw0
-                    blasfeo_dgecpsc(nx, nx+nu, -1.0, dK_dxu_ss, ns * nx + ii * nz, 0,
-                                                                 &dxkzu_dw0, 2 * nx, 0);
+//                    blasfeo_dgecpsc(nx, nx+nu, -1.0, dK_dxu_ss, ns*nx+ii*nz, 0, &dxkzu_dw0, 2*nx, 0); // NOTE(giaf): I think the first argument is wrong, it should be nz !
+                    blasfeo_dgecpsc(nz, nx+nu, -1.0, dK_dxu_ss, ns*nx+ii*nz, 0, &dxkzu_dw0, 2*nx, 0);
```
Comment (giaf) [inline acados/sim/sim_irk_integrator.c]: @FreyJo do you agree with the fix? (nz v.s. nx as first argument)

**Comment by FreyJo** [comment]
For ERK, I am currently not aware how the Hessian is stored.
For IRK, I was also thinking about this at some point. Multiplying with CasADI was easier to implement and potentially CasADi performs similarly, since it can exploit some structural zeros.
Sure, lets discuss this when I am back in syscop office :)

**Comment by giaf** [comment]
@FreyJo for the ERK, I think the compressed format comes form Rien, you could ask not that you are there ;)

**Comment by giaf** [comment]
@FreyJo @zanellia if you have any comment on the PR let me know soon, otherwise I will merge it soon

**Comment by zanellia** [comment]
I would have a look next week. Are we in a huge rush? :P

**Comment by giaf** [comment]
Well, Axel asked me about more than one month ago... now that the project keeps running, it's time to deliver ;)

**Comment by giaf** [comment]
I didn't start doing exact Hessian for fun ;)

**Comment by FreyJo** [comment]
great work!
I looked at this yesterday and just clarified some stuff with @giaf 

The multiplications in the hessian in IRK are now done in blasfeo and are more efficient there compared to CasADi.
This still should be documented in the convention external function file, otherwise I have no remarks.

File: acados/ocp_nlp/ocp_nlp_cost_ls.c [hunk 2]
```diff
@@ -689,25 +712,24 @@ void ocp_nlp_cost_ls_update_qp_matrices(void *config_, void *dims_,
...
                 0, 0, &work->tmp_nz, 0, 1.0, &work->y_ref_tilde, 0, &work->y_ref_tilde, 0);
 
-        blasfeo_dtrmm_rlnn(nu + nx, ny, 1.0, &memory->W_chol,
-                0, 0, &work->Cyt_tilde, 0, 0, &work->tmp_nv_ny,
-                           0, 0);
+        blasfeo_dtrmm_rlnn(nu + nx, ny, 1.0, &memory->W_chol, 0, 0, &work->Cyt_tilde, 0, 0, &work->tmp_nv_ny, 0, 0);
 
         // add hessian of the cost contribution
-        blasfeo_dsyrk_ln(nu + nx, ny, 1.0, &work->tmp_nv_ny, 0, 0, &work->tmp_nv_ny, 0, 0, 1.0,
+        blasfeo_dsyrk_ln(nu + nx, ny, model->scaling, &work->tmp_nv_ny, 0, 0, &work->tmp_nv_ny, 0, 0, 1.0,
```
Comment (zanellia) [inline acados/ocp_nlp/ocp_nlp_cost_ls.c]: Isn't this changing the definition of the Lagrange term from discrete to continuous time?

File: acados/ocp_nlp/ocp_nlp_cost_nls.c [hunk 3]
```diff
@@ -68,6 +68,8 @@ void ocp_nlp_cost_nls_dims_initialize(void *config_, void *dims_, int nx, int nu
     dims->ny = ny;
     dims->ns = ns;
 
+    // TODO nz
```
Comment (zanellia) [inline acados/ocp_nlp/ocp_nlp_cost_nls.c]: Why aren't we setting nz?

File: acados/ocp_nlp/ocp_nlp_cost_ls.c (hunk 2 above)
Comment (giaf) [reply acados/ocp_nlp/ocp_nlp_cost_ls.c]: Nope, it just gives the possibility to optionally introduce a scaling. If model->scaling==1 (which is the default), nothing changes.

But scaling allows e.g. the implementation of Euler integration of continuous costs in the (continuous time) matlab interface.

File: acados/ocp_nlp/ocp_nlp_cost_nls.c (hunk 3 above)
Comment (giaf) [reply acados/ocp_nlp/ocp_nlp_cost_nls.c]: Well, you implemented algebraic variables only for linear least squares, no?

Comment (giaf) [reply acados/ocp_nlp/ocp_nlp_cost_nls.c]: For nonlinear least squares, it is TODO

File: acados/ocp_nlp/ocp_nlp_cost_ls.c [hunk 4]
```diff
@@ -409,6 +410,27 @@ void ocp_nlp_cost_ls_opts_update(void *config_, void *dims_, void *opts_)
...
 
 
+void ocp_nlp_cost_ls_opts_set(void *config_, void *opts_, const char *field, void* value)
+{
+    ocp_nlp_cost_config *config = config_;
+    ocp_nlp_cost_ls_opts *opts = opts_;
+
+    if(!strcmp(field, "exact_hess"))
+    {
+        // do nothing: the exact hessian is always computed
```
[the remark this answers is not in the document]
Comment (giaf) [reply acados/ocp_nlp/ocp_nlp_cost_ls.c]: Good point, I will fix it!

[the remark this answers is not in the document]
Comment (giaf) [reply acados/ocp_nlp/ocp_nlp_cost_ls.c]: Mmm, maybe we should discuss about how this would look like (after this PR has been merged)

**Pull request merged**